### PR TITLE
Use PMIx_log() for show_help()

### PR DESCRIPTION
**Pull request by awlauria**
This requires some changes in prrte for it to work, will post that shortly if this looks ok.

[Refactor show_help() to use the PMIx_Log() api.](https://github.com/openpmix/openpmix/commit/e2e7d1b44b6c69809c861b09439741a9a172dc90) 

- Port over the de-dupping code from prrte.
- Add some additional keys to support a request
  to de-dup logging messages.
- Remove locks from the high level PMIx_Log() api's - they
  don't seem necessary, and make it very easy to deadlock
  when the current thread already holds the global lock.
- Likewise - use PMIx_Log_nb() instead of PMIx_Log() - the
  non-blocking counterpart will allow libevent to trigger the
  callbacks on its own time. Otherwise, in the case of a
  show_help() message from prted that happens before the main thread
  hits the event loop, it will hang forever.
- We can avoid deadlock in show_help calls made prior to
  completing client/server/tool "init" by simply leaving
  the show_help system disabled until we get far enough
  along, and then explicitly "enabling" it. We also never
  call a PMIx API from inside the code base as that can
  lead to loopbacks that deadlock.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>
Signed-off-by: Ralph Castain <rhc@pmix.org>

Co-authored-by: Ralph Castain <rhc@pmix.org>

### CODE REVIEW

**Comment by rhc54** [comment]
> With the change to use PMIx_Log() for show_help(),
> I noticed multiple threads are modifying the iof lists
> at the same time, and would crash with a list left in
> a bad state. This fixes the problem by introducing
> a 'safe' list, which is thread-safe with regards to adding
> and removing objects from them.

Hmmm...no, that isn't how we handle thread safety in this project. We've been burned too many times by threadlock. If multiple threads are touching the iof list, then that is an error in the code - all contact with globals such as the IOF list must be done from inside an event. The correct fix is to ensure that strategy is followed.

I'm not sure where this threading issue is occurring - I gather it is due to the changes in show_help included here? In that case, you probably need to have show_help do a thread-shift before attaching things to the IOF lists - or just call the IOF API functions as they will take care of it for you.

**Comment by awlauria** [comment]
Note 

> > With the change to use PMIx_Log() for show_help(),
> > I noticed multiple threads are modifying the iof lists
> > at the same time, and would crash with a list left in
> > a bad state. This fixes the problem by introducing
> > a 'safe' list, which is thread-safe with regards to adding
> > and removing objects from them.
> 
> Hmmm...no, that isn't how we handle thread safety in this project. We've been burned too many times by threadlock. If multiple threads are touching the iof list, then that is an error in the code - all contact with globals such as the IOF list must be done from inside an event. The correct fix is to ensure that strategy is followed.
> 
> I'm not sure where this threading issue is occurring - I gather it is due to the changes in show_help included here? In that case, you probably need to have show_help do a thread-shift before attaching things to the IOF lists - or just call the IOF API functions as they will take care of it for you.

I don't think it's directly from the changes here, more of a side effect. I didn't touch the IOF lists directly other than adding the thread-safe list struct, but that is easily ripped out - I purposely put all of those changes in a separate commit just in case.

Here's an example backtrack on a coredump of `prterun`:

```
(gdb) info threads
  Id   Target Id                           Frame 
  1    Thread 0x20000960efd0 (LWP 3381632) 0x0000200000b73708 in raise () from /lib64/power9/libc.so.6
* 2    Thread 0x20000005a680 (LWP 3381622) 0x0000200000bc8c58 in __lll_lock_wait_private () from /lib64/power9/libc.so.6
  3    Thread 0x200009a7efd0 (LWP 3381633) 0x0000200000c52968 in select () from /lib64/power9/libc.so.6
  4    Thread 0x200009e8efd0 (LWP 3381634) 0x0000200000c52968 in select () from /lib64/power9/libc.so.6
  5    Thread 0x2000091cefd0 (LWP 3381631) 0x0000200000c6280c in accept4 () from /lib64/power9/libc.so.6
  6    Thread 0x20000abbefd0 (LWP 3381636) 0x0000200000c6068c in epoll_wait () from /lib64/power9/libc.so.6
  7    Thread 0x20000bbfefd0 (LWP 3381640) 0x0000200000c6068c in epoll_wait () from /lib64/power9/libc.so.6
  8    Thread 0x20000b7eefd0 (LWP 3381639) 0x0000200000c6068c in epoll_wait () from /lib64/power9/libc.so.6
  9    Thread 0x20000b3defd0 (LWP 3381638) 0x0000200000c6068c in epoll_wait () from /lib64/power9/libc.so.6
  10   Thread 0x20000afcefd0 (LWP 3381637) 0x0000200000c6068c in epoll_wait () from /lib64/power9/libc.so.6
(gdb) 
```

Thread 1:
```(gdb) thread 1
[Switching to thread 1 (Thread 0x20000960efd0 (LWP 3381632))]
#0  0x0000200000b73708 in raise () from /lib64/power9/libc.so.6
(gdb) bt
#0  0x0000200000b73708 in raise () from /lib64/power9/libc.so.6
#1  0x0000200000b53cc4 in abort () from /lib64/power9/libc.so.6
#2  0x0000200000bbf15c in __libc_message () from /lib64/power9/libc.so.6
#3  0x0000200000bc9fc8 in malloc_printerr () from /lib64/power9/libc.so.6
#4  0x0000200000bcc85c in _int_free () from /lib64/power9/libc.so.6
#5  0x00002000004a2cbc in pmix_iof_write_handler (sd=2, args=4, cbdata=0x2000006d21e8 <pmix_client_globals+1528>) at common/pmix_iof.c:1586
#6  0x000020000076a320 in event_process_active_single_queue () from /lib64/libevent_core-2.1.so.6
#7  0x000020000076b020 in event_base_loop () from /lib64/libevent_core-2.1.so.6
#8  0x00002000004597cc in progress_engine (obj=0x1cf56b20) at runtime/pmix_progress_threads.c:228
#9  0x0000200000ae87c8 in start_thread () from /lib64/power9/libpthread.so.0
#10 0x0000200000c601b8 in clone () from /lib64/power9/libc.so.6
(gdb) 
```
thread 2:
```(gdb) bt
#0  0x0000200000bc8c58 in __lll_lock_wait_private () from /lib64/power9/libc.so.6
#1  0x0000200000bcff98 in malloc () from /lib64/power9/libc.so.6
#2  0x0000200000491c28 in pmix_obj_new_tma (cls=0x2000006d4698 <pmix_iof_write_output_t_class>, tma=0x0)
    at /smpi_dev/awlauria/ompi_builds/main/3rd-party/openpmix/src/class/pmix_object.h:672
#3  0x00002000004918cc in pmix_obj_new_debug_tma (type=0x2000006d4698 <pmix_iof_write_output_t_class>, tma=0x0, 
    file=0x200000641ef0 "common/pmix_iof.c", line=1074)
    at /smpi_dev/awlauria/ompi_builds/main/3rd-party/openpmix/src/class/pmix_object.h:404
#4  0x00002000004a08d4 in write_output_line (name=0x1cf55ea0, channel=0x2000006d2398 <pmix_client_globals+1960>, myflags=0x7fffd921a5c0, 
    stream=4, copystdout=false, copystderr=false, bo=0x7fffd921a5b0) at common/pmix_iof.c:1074
#5  0x00002000004a2464 in pmix_iof_write_output (name=0x1cf55ea0, stream=4, bo=0x7fffd921a660) at common/pmix_iof.c:1483
#6  0x00002000005b74c0 in mylog (source=0x1cf55ea0, data=0x1d064200, ndata=1, directives=0x1d06a840, ndirs=3, 
    cbfunc=0x2000005b3514 <localcbfunc>, cbdata=0x1d0f5aa0) at plog_stdfd.c:108
#7  0x00002000005b4a84 in pmix_plog_base_log (source=0x1cf55ea0, data=0x1d064200, ndata=1, directives=0x1d06a840, ndirs=3, 
    cbfunc=0x20000047db2c <opcbfunc>, cbdata=0x7fffd921abf8) at base/plog_base_stubs.c:200
#8  0x000020000048108c in PMIx_Log_nb (data=0x1d064200, ndata=1, directives=0x1d06a840, ndirs=3, cbfunc=0x20000047db2c <opcbfunc>, 
    cbdata=0x7fffd921abf8) at common/pmix_log.c:256
#9  0x000020000047e050 in PMIx_Log (data=0x1d064200, ndata=1, directives=0x1d06a840, ndirs=3) at common/pmix_log.c:83
#10 0x00002000000ec088 in pmix_server_log (status=0, sender=0x1d062f58, buffer=0x1d063068, tg=65, cbdata=0x0)
    at prted/pmix/pmix_server.c:1462
#11 0x0000200000131df8 in prte_rml_base_process_msg (fd=-1, flags=4, cbdata=0x1d062e30) at rml/rml_base_msg_handlers.c:202
#12 0x000020000076a320 in event_process_active_single_queue () from /lib64/libevent_core-2.1.so.6
#13 0x000020000076b020 in event_base_loop () from /lib64/libevent_core-2.1.so.6
#14 0x000000001000bca4 in main (argc=6, argv=0x7fffd921d458) at prte.c:1186
```
None of the other threads are doing anything interesting.

it looks like thread 2 is directly calling `pmix_iof_write_output()` from `mylog()`. Should it instead be thread-shifted at that point ?

**Comment by rhc54** [comment]
The call to `PMIx_Log_nb` has this note in it:
```c
        /* call down to process the request - the various components
         * will thread shift as required */
        rc = pmix_plog.log(source, data, ndata, cd->directives, cd->ndirs, localcbfunc, cd);
```
The problem here is that the `stdfd` component is violating that directive - it needs to threadshift before calling the iof write function as that function touches global lists/variables.

**Comment by rhc54** [comment]
Actually, after thinking about it some more, let's make the `PMIx_Log_nb` function conform to what we do for all other APIs. If we are not a server, then go ahead and pack/send the message to our server - nothing there touches globals. Otherwise, thread shift and then call the framework API. You'll need to do a quick check to ensure none of the components threadshift, but I suspect they do not.

This will make the code base consistent and easier to maintain going forward. Hopefully, you can just use an existing caddy for the threadshift so this won't be too much trouble.

Appreciate you taking this on!

**Comment by awlauria** [comment]
No problem - I made that change to threadshift and removed the 'safe_list'. I modeled it after what query does, let me know if I missed something or did it wrong.

File: src/common/pmix_log.c
Comment (rhc54) [inline src/common/pmix_log.c:140]: We already did the thread shift - probably should delete this from the comment to avoid future confusion

**Review by rhc54 — CHANGES_REQUESTED** [review]
Just a few cleanups. In general, this looks quite good!

For clarity, let's be sure we understand how this works. Individual processes will call "show_help" which will in turn be funneled into `PMIx_Log_nb`. The plog components in those components will forward the tagged output to their PMIx server, which will hand them up to their host (call it PRRTE for now).

PRRTE daemons will forward the message to the DVM master, which in turn hands them back down to `PMIx_Log`. The DVM master is registered as a PMIX "gateway", and so the plog components know that the message should now be locally displayed. It is at that point that we need to check for duplicates and properly handle them.

I can't tell from all these changes if that logic has been preserved. I don't see anything that would change it, but just wanted to lay it out to ensure you kept it in mind as you have a better mental picture of how this changed code operates.

I guess the only other question is: do we want the "do not aggregate" MCA param? My instinct is "no" as it is up to the host runtime to make that decision. My reasoning is based on the DVM operation, where someone might want to control aggregation on a per-job basis. The MCA param approach forces all jobs to handle it the same way.

So what we want is for the initial source of the message to _always_ tag the message as a show-help message, and for the messages to always come back to the gateway. We want the gateway host to register the job and include the "aggregate show_help" flag in the job-level info (if the host wants to have a param that sets the default behavior, that's their call). The PMIx server needs to look that flag up for the job of the message source when it gets a "show-help" message, and then do the right thing with it. If the flag isn't found (or the flag value is "true"), then we aggregate. If the flag is found and the value is "false", then we don't aggregate. The `PMIX_INFO_TRUE` macro helps with evaluating the flag.

Hope that makes sense - happy to meet on a call to clarify if necessary.

File: src/common/pmix_log.c
Comment (awlauria) [reply src/common/pmix_log.c:72]: Hmm, that complicates things.

If prted or prte (or a call PMIx_foo()) calls show_help while already holding the global lock, it will deadlock. Perhaps we need an internal-only log function for show-help to use that doesn't lock?

**Comment by awlauria** [comment]
> I can't tell from all these changes if that logic has been preserved. I don't see anything that would change it, but just wanted to lay it out to ensure you kept it in mind as you have a better mental picture of how this changed code operates.

Yup - that is how it flows in my head as well.

> I guess the only other question is: do we want the "do not aggregate" MCA param? My instinct is "no" as it is up to the host runtime to make that decision. My reasoning is based on the DVM operation, where someone might want to control aggregation on a per-job basis. The MCA param approach forces all jobs to handle it the same way.
>
> So what we want is for the initial source of the message to always tag the message as a show-help message, and for the messages to always come back to the gateway. We want the gateway host to register the job and include the "aggregate show_help" flag in the job-level info (if the host wants to have a param that sets the default behavior, that's their call). The PMIx server needs to look that flag up for the job of the message source when it gets a "show-help" message, and then do the right thing with it. If the flag isn't found (or the flag value is "true"), then we aggregate. If the flag is found and the value is "false", then we don't aggregate. The PMIX_INFO_TRUE macro helps with evaluating the flag.

I am fine with this approach. I guess I am just not quite sure how/where that logic gets added. Can you point me to an example of where these gets added to the "job-level info"? How is that passed in via `prterun` or `prun` ?

File: src/common/pmix_log.c
Comment (awlauria) [reply src/common/pmix_log.c:184]: I meant to clean this up, happened while I was refactoring code and did this quickly to get it to compile..will fix.

File: src/common/pmix_log.c
Comment (awlauria) [reply src/common/pmix_log.c:72]: Hm, maybe we can add these checks to the thread shift call? that should be ok?

**Comment by rhc54** [comment]
> Can you point me to an example of where these gets added to the "job-level info"? How is that passed in via prterun or prun ?

I was starting to write that all out, but truly it gets a little complicated as it touches several places in PRRTE. Why don't I instead add a commit to this PR that is needed, and then add a commit to your PRRTE one that finishes it over there? I think that will not only complete it, but will help illustrate how you handle such things for the future.

Sound okay?

**Comment by awlauria** [comment]
Sure, if that isn't too much trouble. I'll try to adjust the locking over here.

File: src/common/pmix_log.c
Comment (rhc54) [reply src/common/pmix_log.c:72]: That should never happen. We only hold the lock to check that init counter and then immediately release it. Even if two app threads call PMIx functions, we should never deadlock due to that check.

Comment (rhc54) [reply src/common/pmix_log.c:72]: Let me offer an alternative. The intent of the lock/check was to ensure that the client/server/tool "init" functions had been called and completed prior to allowing someone to call a PMIx API. Otherwise, the caller could begin accessing parts of PMIx before they had been initialized. At the time we did it, there was no problem calling any internal PMIx function from inside the lock because they don't check that lock - it was only done at the beginning of an API. Only exception was in the "init" functions themselves as we had to ensure they completed prior to letting someone into an API.

An aside: this is precisely why we don't lock at the object level. If I am using an object and call a function that calls a function that touches that object, then we deadlock. Very hard to ensure it doesn't happen.

The problem here, therefore, is that we are now calling functions during those "init" routines while holding the lock that could call "show_help", and (as you noted) "show_help" now checks the lock because it calls a PMIx API. Since this is specific to the "init" routines, we have a couple of choices:

* release the lock prior to calling an internal function. We might want to reorder things so that such calls occur late in the "init" routine, after we have setup all the global structures.
* remove the lock and simply put it on the user by dictating that "you cannot call a PMIx function until _after_ the respective init function completes".

I'd personally opt for the first as it is the least risky. It still opens a window for potential trouble, but if we can ensure all globals have been at least initialized, anyone calling an API should at least not segfault.

Make sense?

Comment (awlauria) [reply src/common/pmix_log.c:72]: I think option 1 would require a large refactor, and will be error/mistake prone. Unless you think it will be a smaller change than what I am thinking - modifying the init routines just to accommodate show_help() for error cases seems too risky for me for release branches at this point in the game for v4.2. It could be something considered for master, but I'd be hesitant to bring that to the v4.2 branch (unless, again, it is a smaller refactor than I am thinking).

Option 2 seems __ok__ for now - since as you said, users should not be calling Log before Init anyway. If I had to choose between the two for release branches, I'd probably choose that one.

A third option would be to use a [pthread_readwrite_lock](https://pubs.opengroup.org/onlinepubs/009604599/functions/pthread_rwlock_rdlock.html)(). This will allow readers of these global values (which what PMIx_Log/_nb are doing - just checking the init counters and such) to read these values, even if they hold the write lock. IE PMIx_init() will acquire the write lock in its init routine, and won't deadlock if it calls show_help() because that will only require a read_lock(). This seems like it solves both problems, init functions can safely call lock while still checking these values, and users will still get the error message/check of calling these routines before initializing PMIx. 

I'm not sure if that is safer than modifying/shuffling the init code, but I do think it will be less prone to code drift in the future in avoiding deadlock cases with this path.

Comment (rhc54) [reply src/common/pmix_log.c:72]: let me poke at this a bit over the weekend - we've encountered this a bit before, and i think i recall how we got around it.

**Comment by rhc54** [comment]
@awlauria I pushed a commit to this branch (I wasn't able to come up with a good way to give you a PR - sorry for the "push") that addressed the initialization and aggregation questions. Feel free to squash, edit or revert as you see fit. We should at least "squash" before merging, I think.

**Review by rhc54 — APPROVED** [review]

**Comment by awlauria** [comment]
Thanks! Will take a look and squash it down in a few minutes.

**Comment by awlauria** [comment]
@rhc54 lgtm, thanks! I squashed it down. I did add one minor change to free the info's in show_help:

https://github.com/openpmix/openpmix/compare/0a104f9b0b3b3789b5abb3c4bea95f55e74501eb..71468678f29326c83c4846a76d33c18aae67d0d1

**Comment by rhc54** [comment]
> @rhc54 lgtm, thanks! I squashed it down. I did add one minor change to free the info's in show_help:
> 
> https://github.com/openpmix/openpmix/compare/0a104f9b0b3b3789b5abb3c4bea95f55e74501eb..71468678f29326c83c4846a76d33c18aae67d0d1

Nice catch - thx!

**Comment by awlauria** [comment]
@rhc54 thoughts on adding a wrapper function that prte/opal can call to handle this thread-shift? Basically prte_show_help/opal_show_help will pass the infos/ndirs down to pmix for processing. Something like this:

```
+void pmix_show_help_caller(pmix_info_t *info, size_t ninfos,
+                           pmix_info_t *dir, size_t ndirs) {
+
+    pmix_shift_caddy_t *cd;
+
+    cd = PMIX_NEW(pmix_shift_caddy_t);
+    cd->ninfo = ninfos;
+    cd->info = info;
+
+    cd->ndirs = ndirs;
+    cd->directives = dir;
+    cd->cbfunc.opcbfn = show_help_cbfunc;
+    cd->cbdata = cd;
+    cd->proc = NULL;
+    PMIX_THREADSHIFT(cd, pmix_log_local_op);
+
+}
+
 static void local_delivery(const char *file, const char *topic, char *msg)
 {
-    pmix_shift_caddy_t *cd;
+
+    pmix_info_t *info, *dirs;
+    size_t ninfos = 1, ndirs = 2;
 
     if (!pmix_show_help_enabled) {
         /* the show help subsystem has not yet been enabled,
@@ -80,18 +100,14 @@ static void local_delivery(const char *file, const char *topic, char *msg)
         return;
     }
 
-    cd = PMIX_NEW(pmix_shift_caddy_t);
-    cd->ninfo = 1;
-    PMIX_INFO_CREATE(cd->info, cd->ninfo);
-    PMIX_INFO_LOAD(&cd->info[0], PMIX_LOG_STDERR, msg, PMIX_STRING);
-    cd->ndirs = 2;
-    PMIX_INFO_CREATE(cd->directives, cd->ndirs);
-    PMIX_INFO_LOAD(&cd->directives[0], PMIX_LOG_KEY, file, PMIX_STRING);
-    PMIX_INFO_LOAD(&cd->directives[1], PMIX_LOG_VAL, topic, PMIX_STRING);
-    cd->cbfunc.opcbfn = show_help_cbfunc;
-    cd->cbdata = cd;
-    cd->proc = NULL;
-    PMIX_THREADSHIFT(cd, pmix_log_local_op);
+    PMIX_INFO_CREATE(info, ninfos);
+    PMIX_INFO_LOAD(&info[0], PMIX_LOG_STDERR, msg, PMIX_STRING);
+
+    PMIX_INFO_CREATE(dirs, ndirs);
+    PMIX_INFO_LOAD(&dirs[0], PMIX_LOG_KEY, file, PMIX_STRING);
+    PMIX_INFO_LOAD(&dirs[1], PMIX_LOG_VAL, topic, PMIX_STRING);
+
+    pmix_show_help_caller(info, ninfos, dirs, ndirs);
 }
```

it'll reduce code duplication. A lot more code de-duplication can be done - I don't think there's any reason why `prte|opal_show_help` can't be completely replaced with `pmix_show_help`, either.

**Comment by rhc54** [comment]
> it'll reduce code duplication. A lot more code de-duplication can be done - I don't think there's any reason why `prte|opal_show_help` can't be completely replaced with `pmix_show_help`, either.

I'm doing that for `prte` right now 😄

**Comment by awlauria** [comment]
Hmm, it's actually easier than I thought. for prte all that we need to port over to pmix is prte_show_help_norender(). after that it's a simple code change from prte -> pmix. Will append to my pr.

**Comment by awlauria** [comment]
@rhc54 ah - well I appended it here: https://github.com/openpmix/prrte/pull/1350

if you'd like to do it differently or you're already far along that's fine and I'll remove it. It requires a small change to port `norender` over here, but that's it.

**Pull request merged**
